### PR TITLE
fix: use void return type for command implementations

### LINKED ISSUE #18915: Argument of type '() => void' is not assignable to parameter of type '() => Chainable<any>'.
### Current behavior

With the upgrade to v9, Cypress is reporting the following error when adding a custom method:
```
Argument of type '() => void' is not assignable to parameter of type '() => Chainable<any>'.
```

### Desired behavior

According to the documentation, no other code should be needed to make this work.

Thus I believe that void should be a valid type to be supported.

### Test code to reproduce

```
Cypress.Commands.add('login', () => {
  // do something
})
```

### Cypress Version

9.0.0

### Other

_No response_

### PR DESCRIPTION
**Pull request by remcohaszing**
- Closes #18915

### User facing changelog
The TypeScript return type of Cypress command implementations has been fixed.

### Additional details
The `ConditionalKeys` type has been copied from `type-fest`, because I wasn’t sure if you want me to add it as a dependency.

### How has the user experience changed?
Custom Cypress command implementations no longer show a TypeScript error when they use a chainable return value.

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue or this PR been tagged with a release in ZenHub? 
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [x] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [ ] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

**Review by chrisbreiding — CHANGES_REQUESTED** [review]
This seems to allow any return value from a custom command implementation.

9.0.0:

Screen Shot 2021-11-22 at 9 30 14 AM

This PR:

Screen Shot 2021-11-22 at 9 31 51 AM

**Comment by remcohaszing** [comment]
@chrisbreiding:

> This seems to allow any return value from a custom command implementation.

That’s correct. Based on feedback from #18915 that is the desired behaviour, because the return value is discarded. However, your comment implies this is incorrect. Could you verify this, and if so, explain what is allowed to be returned?

**Comment by BlueWinds** [comment]
> @chrisbreiding:
> 
> > This seems to allow any return value from a custom command implementation.
> 
> That’s correct. Based on feedback from #18915 that is the desired behaviour, because the return value is discarded. However, your comment implies this is incorrect. Could you verify this, and if so, explain what is allowed to be returned?

Agreed, allowing any return value seems correct to me. Consider the following:
```
it('works', () => {
  Cypress.Commands.add('returnTrue', () => true)
  cy.returnTrue().should('eq', true)
})
```

This is perfectly valid and supported, and our types should reflect that. We wrap the function the user passes in in order to ensure the chainability - which means that the function the user passes to `Commands.add` does not itself need to return a chainable value.

**Comment by chrisbreiding** [comment]
> which means that the function the user passes to Commands.add does not itself need to return a chainable value

I agree with that. But it does weaken the type safety a lot, doesn't it? The type can say the custom command should return an element, while the custom command itself returns a string, and that passes the type-checking.

**Comment by chrisbreiding** [comment]
> the actual behaviour can’t be described using TypeScript

That was my hunch, but I guess I was hoping there was some way to express the types correctly with maximal type safety. I'm far from a TypeScript expert, so I certainly defer to your knowledge on the matter.

> It’s better to have less strict types that are correct than super strict incorrect types that make get in the way of using the code correctly.

Definitely agree.

> So, if I understand correctly:
> 
>     commands always return a chainable when used
>     command implementations may return a chainable
>     command implementations may return a value that will be wrapped in a chainable
>     command implementations may return undefined

Yes, I think that sums it up nicely.

---

Sorry, I should have just asked some questions instead of submitting a review. I was mostly trying to understand the implications of this. Now that I understand the issue better, it seems like this is the best we can do to add correct types for custom commands.

File: cli/types/cypress-type-helpers.d.ts [hunk 1]
```diff
@@ -1,2 +1,19 @@
 // type helpers
 type Nullable<T> = T | null
+
+// https://github.com/sindresorhus/type-fest/blob/c8ee2d262fa970ed358e2f07b9ecc06524cac2be/source/conditional-keys.d.ts#L31-L45
+type ConditionalKeys<Base, Condition> = NonNullable<
+	// Wrap in `NonNullable` to strip away the `undefined` type from the produced union.
```
Comment (chrisbreiding) [inline cli/types/cypress-type-helpers.d.ts:6]: Could the tabs in this file be converted to spaces, for the sake of consistency with the rest of the codebase?

**Review by chrisbreiding — CHANGES_REQUESTED** [review]

**Comment by remcohaszing** [comment]
If I change the return type to `undefined`, I’m enforced to return explicitly, which is undesired. I think it’s best to keep this as-is.

I just want to add that the `void` type can be changed to `unknown` or `any` as well. These all yield the same result.

**Review by chrisbreiding — APPROVED** [review]

**Review by emilyrohrbough — APPROVED** [review]

**Comment by chrisbreiding** [comment]
Closing in favor of #19003 as [suggested here](https://github.com/cypress-io/cypress/pull/19003#issuecomment-979054587).

**Pull request closed without merging**